PackManagement.php

```diff
@@ -504,6 +776,48 @@ public function dbReleaseLock()
 
 
 
+	// Delete a pack.
+	public function deletePack( $projectID, $catID, $packID )
+	{
+		$sql = 'UPDATE redcap_external_module_settings ems SET ems.value = JSON_REMOVE(ems.value' .
+		       ',REPLACE(JSON_UNQUOTE(JSON_SEARCH(ems.value,\'one\',?,NULL,\'$[*].id\')),' .
+		       '\'.id\',\'\')) WHERE ems.external_module_id = (SELECT em.external_module_id ' .
+		        'FROM redcap_external_modules em WHERE em.directory_prefix = ? LIMIT 1) ' .
+		        'AND ems.key = ? AND JSON_CONTAINS( JSON_EXTRACT( ems.value, \'$[*].id\' ), ? )';
+		$sqlParams = [ $packID, $this->getModuleDirectoryBaseName(),
+		               'p' . $projectID . '-packlist-' . $catID, json_encode( $packID ) ];
+		$this->query( $sql, $sqlParams );
+	}
+
+
+
+	// Echo plain text to output (without Psalm taints).
+	// Use only for e.g. JSON or CSV output.
+	function echoText( $text )
+	{
+		$text = htmlspecialchars( $text, ENT_QUOTES | ENT_SUBSTITUTE | ENT_XHTML );
+		$chars = [ '&amp;' => 38, '&quot;' => 34, '&apos;' => 39, '&lt;' => 60, '&gt;' => 62 ];
+		$text = preg_split( '/(&(?>amp|quot|apos|lt|gt);)/', $text, -1, PREG_SPLIT_DELIM_CAPTURE );
+		foreach ( $text as $part )
+		{
+			echo isset( $chars[ $part ] ) ? chr( $chars[ $part ] ) : $part;
+		}
+	}
+
+
+
+	// Escapes text string for inclusion in JavaScript.
+	function escapeJSString( $text )
+	{
+		return '"' . $this->escape( substr( json_encode( (string)$text,
+		                                                 JSON_HEX_QUOT | JSON_HEX_APOS |
```

Comment:
am I getting deja vu with this string of predefined constants like JSON_HEX_QUOT?  Should we have it all done once?

Reply:
Here it's within a function so that's the intention.